crates/lib/src/execution_data.rs

```diff
@@ -1,120 +1,345 @@
+    /// registers would result in a jagged matrix. [`QpuReadout`] data is captured per measure,
+    /// meaning a value is returned for every measure to a memory reference, not just once per shot.
+    /// This is often the case in programs that use mid-circuit measurement or dynamic control flow,
+    /// where measurements to the same memory reference might occur multiple times in a shot, or be
+    /// skipped conditionally. In these cases, building a rectangular [`RegisterMatrix`] would
+    /// necessitate making assumptions about the data that could skew the data in undesirable ways.
+    /// Instead, it's recommended to manually build a matrix from [`QpuReadout`] that accurately
+    /// selects the last value per-shot based on the program that was run.
+    pub fn to_readout_map(&self) -> Result<ReadoutMap, RegisterMatrixConversionError> {
+        match self {
+            ReadoutData::Qvm(data) => ReadoutMap::from_qvm_memory(data),
+            ReadoutData::Qpu(data) => ReadoutMap::from_qpu_readout_data(data),
+        }
+    }
+}
 
 impl ReadoutMap {
-    /// Given a known readout field name and index, return the result's [`ReadoutValues`], if any.
+    /// Returns the [`RegisterMatrix`] for the given register, if it exists.
     #[must_use]
-    pub fn get_readout_values(&self, field: String, index: u64) -> Option<ReadoutValues> {
-        let readout_values = self.0.get(&MemoryReference { name: field, index })?;
-
-        Some(readout_values.clone())
+    pub fn get_register_matrix(&self, register_name: &str) -> Option<&RegisterMatrix> {
+        self.0.get(register_name)
     }
 
-    /// Given a known readout field name, return the result's [`ReadoutValues`] for all indices, if any.
-    pub fn get_readout_values_for_field(
-        &self,
-        field: &str,
-    ) -> Result<Option<Vec<Option<ReadoutValues>>>, TryFromIntError> {
-        let mut readout_values = Vec::new();
-        for (memref, values) in &self.0 {
-            let MemoryReference { name, index } = memref;
-            let index = usize::try_from(*index)?;
-            if name == field {
-                if readout_values.len() <= index {
-                    readout_values.resize(index + 1, None);
-                }
-                readout_values[index] = Some(values.clone());
-            }
-        }
+    /// Returns a [`ReadoutMap`] with the underlying [`RegisterMatrix`] data
+    #[must_use]
+    pub fn from_hashmap(map: HashMap<String, RegisterMatrix>) -> Self {
+        Self(map)
+    }
 
-        Ok((!readout_values.is_empty()).then_some(readout_values))
+    /// Returns a [`ReadoutMap`] built from [`QvmMemory`]
+    pub fn from_qvm_memory(memory: &QvmMemory) -> Result<Self, RegisterMatrixConversionError> {
+        Ok(Self(
+            memory
+                .iter()
+                .map(|(name, register)| {
+                    let register_matrix = match register {
```

Comment:
I really want to suggest a macro to deduplicate the code in each branch, but not sure how much that would help maintainability.

Reply:
Yeah, if we were doing this in more than one place, a could see an argument for a macro. Just using one for this though feels like it might hurt maintainability more than help.